gpiozero/pins/pi.py

```diff
@@ -0,0 +1,251 @@
+        elif set(spi_args) <= set(pin_defaults):
+            spi_args = {
+                key: self._to_gpio(spi_args.get(key, default))
+                for key, default in pin_defaults.items()
+                }
+        elif set(spi_args) <= set(dev_defaults):
+            spi_args = {
+                key: spi_args.get(key, default)
+                for key, default in dev_defaults.items()
+                }
+            if spi_args['port'] != 0:
+                raise SPIBadArgs('port 0 is the only valid SPI port')
+            if spi_args['device'] not in (0, 1):
+                raise SPIBadArgs('device must be 0 or 1')
+            spi_args = {
+                key: value if key != 'select_pin' else (8, 7)[spi_args['device']]
+                for key, value in pin_defaults.items()
+                }
+        else:
+            raise SPIBadArgs(
+                'you must either specify port and device, or clock_pin, '
+                'mosi_pin, miso_pin, and select_pin; combinations of the two '
+                'schemes (e.g. port and clock_pin) are not permitted')
+        return spi_args, kwargs
+
+
+class PiPin(Pin):
+    """
+    Abstract base class representing a multi-function GPIO pin attached to a
+    Raspberry Pi.
+    """
+    def __init__(self, factory, number):
+        super(PiPin, self).__init__()
+        self._factory = proxy(factory)
+        self._when_changed_lock = RLock()
+        self._when_changed = None
+        self._number = number
+        try:
+            factory.pi_info.physical_pin(self.address[-1])
+        except PinNoPins:
+            warnings.warn(
+                PinNonPhysical(
+                    'no physical pins exist for %s' % self.address[-1]))
+
+    @property
+    def number(self):
+        return self._number
+
+    @property
+    def factory(self):
+        return self._factory
+
+    def _get_address(self):
+        return self.factory.address + ('GPIO%d' % self.number,)
+
+    def _call_when_changed(self):
+        method = self.when_changed()
+        if method is None:
+            self.when_changed = None
+        else:
+            method()
+
+    def _get_when_changed(self):
+        return self._when_changed
+
+    def _set_when_changed(self, value):
+        # Have to take care, if value is either a closure or a bound method,
+        # not to keep a strong reference to the containing object
+        with self._when_changed_lock:
+            if self._when_changed is None and value is not None:
+                if isinstance(value, MethodType):
+                    self._when_changed = WeakMethod(value)
+                else:
+                    self._when_changed = ref(value)
+                self._enable_event_detect()
+            elif self._when_changed is not None and value is None:
+                self._disable_event_detect()
+                self._when_changed = None
+            elif value is None:
+                self._when_changed = None
```

Comment:
Yeah - I need to tidy that up a bit; it used to have a nice "when setting", "when unsetting", "when changing" branch structure and now it's all messy. I'll have a crack at that in a mo.